File: truco_bot/core/rules.py
from truco_bot.core.card import Card


def calculate_envido(cards: list[Card]) -> int:
    by_suit: dict[int, list[int]] = {}
    for card in cards:
        by_suit.setdefault(card.suit.value, []).append(card.envido_value)
    
    max_envido = 0
    for values in by_suit.values():
        if len(values) >= 2:
            values.sort(reverse=True)
            score = values[0] + values[1] + 20
        else:
            score = values[0]
        max_envido = max(max_envido, score)
        
    return max_envido

def compare_cards(c1: Card, c2: Card) -> int:
    # return 1, -1, or 0 based on rank directly.
    if c1.truco_rank > c2.truco_rank:
        return 1
    if c1.truco_rank < c2.truco_rank:
        return -1
    return 0

def resolve_hand(trick_results: list[int], mano: int = 0) -> int | None:
    # trick_results: 1 (P0 win), -1 (P1 win), 0 (parda)
    p0 = trick_results.count(1)
    p1 = trick_results.count(-1)
    
    # 2 trick wins immediately
    if p0 >= 2: return 0
    if p1 >= 2: return 1
    
    if len(trick_results) >= 2:
        # First trick decided, second parda -> first winner wins
        if trick_results[0] != 0 and trick_results[1] == 0:
            return 0 if trick_results[0] == 1 else 1
        # First trick parda, second decided -> second winner wins
        if trick_results[0] == 0 and trick_results[1] != 0:
            return 0 if trick_results[1] == 1 else 1

    if len(trick_results) == 3:
        # Third trick parda -> first trick winner wins
        if trick_results[2] == 0 and trick_results[0] != 0:
            return 0 if trick_results[0] == 1 else 1
        # T1 parda, T2 parda, T3 decided -> T3 winner wins
        if trick_results[0] == 0 and trick_results[1] == 0 and trick_results[2] != 0:
            return 0 if trick_results[2] == 1 else 1
        # All parda -> mano wins
        if trick_results.count(0) == 3:
            return mano
            
    return None

def calculate_falta_envido_points(
            score_p0: int, 
            score_p1: int, 
            max_score: int = 30
        ) -> int:
    return max_score - max(score_p0, score_p1)

File: crates/truco_engine/src/rules.rs
//! Game rules: envido calculation, hand resolution, and falta envido point determination.

use crate::card::{CARD_ENVIDO, CARD_SUITS};

pub fn calculate_envido(cards: &[u8]) -> u8 {
    if cards.is_empty() {
        return 0;
    }
    if cards.len() == 1 {
        return CARD_ENVIDO[cards[0] as usize];
    }

    // ponytail: stack-allocated suits buffer for 0..3 cards; zero heap allocation
    let mut suit_cards: [[u8; 3]; 4] = [[0; 3]; 4];
    let mut suit_counts: [usize; 4] = [0; 4];
    let mut max_single = 0u8;

    for &c in cards {
        let s = CARD_SUITS[c as usize] as usize;
        let e = CARD_ENVIDO[c as usize];
        if e > max_single {
            max_single = e;
        }
        suit_cards[s][suit_counts[s]] = e;
        suit_counts[s] += 1;
    }

    let mut max_suited = 0u8;
    for s in 0..4 {
        match suit_counts[s] {
            2 => {
                let val = 20 + suit_cards[s][0] + suit_cards[s][1];
                if val > max_suited {
                    max_suited = val;
                }
            }
            3 => {
                let mut v = [suit_cards[s][0], suit_cards[s][1], suit_cards[s][2]];
                v.sort_unstable();
                let val = 20 + v[1] + v[2];
                if val > max_suited {
                    max_suited = val;
                }
            }
            _ => {}
        }
    }

    if max_suited > 0 {
        max_suited.max(max_single)
    } else {
        max_single
    }
}

pub fn resolve_hand(trick_results: &[i8], mano: u8) -> Option<u8> {
    let p0 = trick_results.iter().filter(|&&r| r == 1).count();
    let p1 = trick_results.iter().filter(|&&r| r == -1).count();

    if p0 >= 2 {
        return Some(0);
    }
    if p1 >= 2 {
        return Some(1);
    }

    if trick_results.len() >= 2 {
        // T1 won, T2 parda -> T1 winner wins
        if trick_results[0] != 0 && trick_results[1] == 0 {
            return Some(if trick_results[0] == 1 { 0 } else { 1 });
        }
        // T1 parda, T2 won -> T2 winner wins
        if trick_results[0] == 0 && trick_results[1] != 0 {
            return Some(if trick_results[1] == 1 { 0 } else { 1 });
        }
    }

    if trick_results.len() == 3 {
        // T3 parda -> T1 winner wins
        if trick_results[2] == 0 && trick_results[0] != 0 {
            return Some(if trick_results[0] == 1 { 0 } else { 1 });
        }
        // T1 parda, T2 parda, T3 decided -> T3 winner wins
        if trick_results[0] == 0 && trick_results[1] == 0 && trick_results[2] != 0 {
            return Some(if trick_results[2] == 1 { 0 } else { 1 });
        }
        // All parda -> mano wins
        if trick_results.iter().filter(|&&r| r == 0).count() == 3 {
            return Some(mano);
        }
    }

    None
}

#[inline(always)]
pub fn calculate_falta_envido_points(score_p0: u8, score_p1: u8, max_score: u8) -> u8 {
    // ponytail: saturating sub prevents underflow when already at max score
    max_score.saturating_sub(score_p0.max(score_p1))
}
